Java/core/src/main/java/com/amazon/randomcutforest/state/store/NodeStoreMapper.java

```diff
@@ -33,34 +37,132 @@
+                nodeFreeIndexes, nodeFreeIndexPointer, leafFreeIndexes, leafFreeIndexPointer);
     }
 
     @Override
     public NodeStoreState toState(NodeStore model) {
         NodeStoreState state = new NodeStoreState();
-        state.setLeftIndex(Arrays.copyOf(model.leftIndex, model.leftIndex.length));
-        state.setRightIndex(Arrays.copyOf(model.rightIndex, model.rightIndex.length));
-        state.setParentIndex(Arrays.copyOf(model.parentIndex, model.parentIndex.length));
-        state.setMass(Arrays.copyOf(model.mass, model.mass.length));
-        state.setCutDimension(Arrays.copyOf(model.cutDimension, model.cutDimension.length));
-        state.setCutValue(Arrays.copyOf(model.cutValue, model.cutValue.length));
-        state.setFreeIndexes(Arrays.copyOf(model.getFreeIndexes(), model.getFreeIndexPointer() + 1));
+        state.setCapacity(model.getCapacity());
+        state.setCompressed(compress);
+        state.setSamplerNeeded(samplerNeeded);
+
+        BitSet bits = new BitSet(2 * model.getCapacity());
+        int[] leftIndex = Arrays.copyOf(model.leftIndex, model.leftIndex.length);
+        int[] rightIndex = Arrays.copyOf(model.rightIndex, model.rightIndex.length);
+        boolean check = compress && feasibleCanonical && reduceToBits(model.size(), leftIndex, rightIndex);
+        state.setCanonicalAndNotALeaf(check);
+        if (check) { // can have a canonical representation saving a lot of space
+            state.setLeftIndex(ArrayPacking.pack(leftIndex, state.isCompressed()));
+            state.setRightIndex(ArrayPacking.pack(rightIndex, state.isCompressed()));
+            state.setSize(model.size());
+        } else { // the temporary array leftIndex and rightIndex may be corrupt in reduceToBits()
+            state.setLeftIndex(ArrayPacking.pack(model.leftIndex, compress));
+            state.setRightIndex(ArrayPacking.pack(model.rightIndex, compress));
+        }
+
+        state.setCutDimension(ArrayPacking.pack(model.cutDimension, compress));
+        state.setCutValueDouble(Arrays.copyOf(model.cutValue, model.cutValue.length));
+
+        state.setNodeFreeIndexes(ArrayPacking.pack(model.getNodeFreeIndexes(), compress));
+        state.setNodeFreeIndexPointer(model.getNodeFreeIndexPointer());
+        state.setLeafFreeIndexes(ArrayPacking.pack(model.getLeafFreeIndexes(), compress));
+        state.setLeafFreeIndexPointer(model.getLeafFreeIndexPointer());
+        if (!samplerNeeded) {
+            state.setLeafPointIndex(ArrayPacking.pack(model.getLeafPointIndex(), compress));
+            state.setLeafmass(ArrayPacking.pack(model.getLeafMass(), compress));
+        }
+
         return state;
     }
+
+    boolean reduceToBits(int size, int[] leftIndex, int[] rightIndex) {
+        boolean check = leftIndex.length == rightIndex.length;
+        int leafCounter = leftIndex.length;
+        int nodeCounter = 1;
+        for (int i = 0; i < size && check; i++) {
+            if (leftIndex[i] != NULL) {
```

Comment:
What if `leftIndex[i] == null` but `rightIndex[i] != null` or if the left index is a leaf but the right one isn't? Should we check those cases?

Reply:
In that first case the tree is malformed in the first place. But the left can be a leaf and the right can be non-leaf.
But the point about validation in nodestore makes sense.